mismi-s3/src/Mismi/S3/Commands.hs

```diff
@@ -466,74 +476,61 @@ listRecursively a =
-  (c, r) <- liftIO $ (,) <$> newChan <*> newChan
   e <- ask
+  bimapEitherT SyncError id .
+    consume (sinkQueue e (listRecursively' source)) fork $ worker source dest mode e
 
-  -- worker
-  tid <- liftIO $ forM [1..fork] (const . forkIO $ worker source dest mode e c r)
+newtype SyncError =
+  SyncError (RunError SyncWorkerError)
 
-  -- sink list to channel
-  let l = listRecursively' source
-  i <- sinkChanWithDelay 50000 l c
+renderSyncError :: SyncError -> Text
+renderSyncError (SyncError r) =
+  renderRunError r renderSyncWorkerError
 
-  -- wait for threads and lift errors
-  r' <- liftIO $ waitForNResults i r
-  liftIO $ forM_ tid killThread
-  forM_ r' hoistWorkerResult
+data SyncWorkerError =
+   SyncInvariant Address Address
+ | OutputExists Address
+ | SyncAws Error
+ deriving Show
 
-grantReadAccess :: Address -> ReadGrant -> AWS ()
-grantReadAccess a g =
-  void . send $ (fencode' P.putObjectACL a & P.poaGrantRead .~ Just (readGrant g))
-
-hoistWorkerResult :: WorkerResult -> AWS ()
-hoistWorkerResult =
-  foldWR throwM (pure ())
-
-worker :: Address -> Address -> SyncMode -> Env -> Chan Address -> Chan WorkerResult -> IO ()
-worker source dest mode e c errs = forever $ do
-  let invariant = pure . WorkerErr $ Invariant "removeCommonPrefix"
-      keep :: Address -> Key -> IO WorkerResult
-      keep a k = (keep' a k >> return WorkerOk) `catch` \er -> return (WorkerErr er)
-
-      keep' :: Address -> Key -> IO ()
-      keep' a k = do
-        let out = withKey (</> k) dest
-            action :: AWS ()
-            action = do
-              let cp = copy a out
-                  ex = exists out
-                  te = throwM $ Target a out
-              foldSyncMode
-                (ifM ex te cp)
-                cp
-                (ifM ex (return ()) cp)
-                mode
-        eitherT throwM pure $ runAWS e action
-
-  a <- readChan c
-  wr <- maybe invariant (keep a) $ removeCommonPrefix source a
-  writeChan errs wr
-
-data WorkerResult =
-    WorkerOk
-  | WorkerErr S3Error
-
-foldWR :: (S3Error -> m a) -> m a -> WorkerResult -> m a
-foldWR e a = \case
-  WorkerOk -> a
-  WorkerErr err -> e err
+renderSyncWorkerError :: SyncWorkerError -> Text
+renderSyncWorkerError w =
+  case w of
+    SyncInvariant a b ->
+      "Remove common prefix invariant: " <>
+      "[" <> addressToText b <> "] is not a common prefix of " <>
+      "[" <> addressToText a <> "]"
+    OutputExists a ->
+      "Can not copy to an address that already exists [" <> addressToText a <> "]"
+    SyncAws e ->
+      "AWS failure during 'sync': " <> renderError e
+
+worker :: Address -> Address -> SyncMode -> Env -> Address -> EitherT SyncWorkerError IO ()
+worker input output mode env f = runAWST env SyncAws $ do
+  n <- maybe (left $ SyncInvariant input f) right $ removeCommonPrefix input f
+  let out = withKey (</> n) output
+      cp = lift $ copy f out
+  foldSyncMode
+    (ifM (lift $ exists out) (left $ OutputExists out) cp)
+    cp
```

Comment:
Mmmm, this looks wrong, isn't this meant to be overwrite? Won't this check and fail?